core/src/main/java/org/mapfish/print/http/HttpRequestCache.java

```diff
@@ -0,0 +1,178 @@
+ * 
+ * @author Niels Charlier
+ *
+ */
+public final class HttpRequestCache {
+    
+    private static final Logger LOGGER = LoggerFactory.getLogger(HttpRequestCache.class);
+    
+    private static class CachedClientHttpResponse extends AbstractClientHttpResponse {
+        
+        private final File cachedFile;
+        private final HttpHeaders headers;
+        private final int status;
+        private final String statusText;
+        
+        public CachedClientHttpResponse(final ClientHttpResponse originalResponse) throws IOException {
+            this.headers = originalResponse.getHeaders();
+            this.status = originalResponse.getRawStatusCode();
+            this.statusText = originalResponse.getStatusText();
+            this.cachedFile = File.createTempFile("cacheduri", null);
+            InputStream is = originalResponse.getBody();
+            try {
+                OutputStream os = new FileOutputStream(this.cachedFile);
+                try {
+                    IOUtils.copy(is, os);
+                } finally {
+                    os.close();
+                }
+            } finally {
+                is.close();
+            }
+        }
+
+        @Override
+        public InputStream getBody() throws IOException {
+            return new FileInputStream(this.cachedFile);
+        }
+
+        @Override
+        public HttpHeaders getHeaders() {
+            return this.headers;
+        }
+
+        @Override
+        public int getRawStatusCode() throws IOException {
+            return this.status;
+        }
+
+        @Override
+        public String getStatusText() throws IOException {
+            return this.statusText;
+        }
+
+        @Override
+        public void close() {
+            
+        }
+    }
+
+    private static class CachedClientHttpRequest implements ClientHttpRequest, Callable<Void> {
+        private final ClientHttpRequest originalRequest;
+        private CachedClientHttpResponse response;
+        
+        public CachedClientHttpRequest(final ClientHttpRequest request) throws IOException {
+            this.originalRequest = request;
+        }
+        
+        @Override
+        public HttpMethod getMethod() {
+            return this.originalRequest.getMethod();
+        }
+
+        @Override
+        public URI getURI() {
+            return this.originalRequest.getURI();
+        }
+
+        @Override
+        public HttpHeaders getHeaders() {
+            return this.originalRequest.getHeaders();
+        }
+
+        @Override
+        public OutputStream getBody() throws IOException {
+            //body should be written before creating this object
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public ClientHttpResponse execute() throws IOException {
+            if (this.response == null) {
+                LOGGER.warn("Attempting to load cached URI before actual caching: " + this.originalRequest.getURI());
```

Comment:
If the request failed, is `this.reponse` also `null`? Is the warning then still appropriate?